File: Problem_2 - 1.py
"""
Time Complexity : O(n^2). 
Space Complexity : O(n^2).
    
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        if not preorder:
            return None 
        
        root = TreeNode(preorder[0])
        rootval = preorder[0] 
        index = 0 
        
        for i in range(len(inorder)):
            if inorder[i] == rootval:
                index = i 
        
        inleft = inorder[0:index]
        inright = inorder[index+1:]
        preleft = preorder[1:index+1]
        preright = preorder[index+1:]
        
        root.left = self.buildTree(preleft, inleft)
        root.right = self.buildTree(preright, inright)
        
        return root
            
            

File: src/Problem2_1.java
//Time Complexity : O(n^2)
//Space Complexity : O(n*h), where h is the height of the tree.
//Code run successfully on LeetCode.

public class Problem2_1 {

    public TreeNode buildTree(int[] preorder, int[] inorder) {
        
        if(preorder == null || preorder.length == 0 || inorder == null || inorder.length == 0)
            return null;
        
        int rootIdx = -1;
        int rootVal = preorder[0];
        
        TreeNode root = new TreeNode(rootVal);
        
        for(int i =0; i < inorder.length; i++){
            if(inorder[i] == rootVal)
            {
                rootIdx = i;
                break;
            }
        }
        
        int[] inorderLeft = Arrays.copyOfRange(inorder, 0, rootIdx);
        int[] inorderRight = Arrays.copyOfRange(inorder, rootIdx +1, inorder.length);
        int[] preorderLeft = Arrays.copyOfRange(preorder, 1, rootIdx+1);
        int[] preorderRight = Arrays.copyOfRange(preorder, rootIdx +1, preorder.length);
        
        root.left = buildTree(preorderLeft, inorderLeft);
        root.right = buildTree(preorderRight, inorderRight);
        
        return root;
    }
}
